### PR TITLE
Recovering PR #706: Supporting native JupyterHub OAuthenticator alongside Auth0 for our hubs

### PR DESCRIPTION
**Pull request by sgibson91**
### Summary

This PR is a reconstruction of PR #706 which failed on merge due to JSON schema validation errors. In addition to reconstructing that, this PR also aims to address the JSON schema validation errors and extends the `validate()` function in `deployer` to also validate the secret config, if it exists.

fixes https://github.com/2i2c-org/pilot-hubs/issues/625

### Changes to `config/hubs/schema.yaml`

I have [borrowed a pattern from @damianavila](https://github.com/2i2c-org/pilot-hubs/blob/master/config/hubs/schema.yaml#L115-L121) to make the `auth0.connection` property conditional on `auth0.enabled` which is now a fully fledged property of `auth0`.

I have also set the default value of `auth0.enabled` to be `true` so that we don't have to go through every `*.cluster.yaml` file and add the `enabled` key for `auth0`.

### CODE REVIEW

**Comment by sgibson91** [comment]
I have deployed this to Pangeo staging (partnered with #707) and it works! So marking this as ready for review :)

File: config/hubs/schema.yaml [hunk 1]
```diff
@@ -161,10 +161,18 @@ properties:
...
+              Most hubs use Auth0 for authentication, and we dynamically fetch the credentials
               needed for each hub - client_id, client_secret, callback_url - on deploy. This
               block contains configuration on how auth0 should be configured for this hub.
             properties:
+              enabled:
+                type: boolean
+                default: true
+                description: |
+                  Whether or not to enable Auth0 authentication for this hub.
+                  This should almost always true. It should only be false if
```
Comment (sgibson91) [inline config/hubs/schema.yaml:173]:
```suggestion
                  This should almost always be true. It should only be false if
```

**Comment by choldgraf** [comment]
Just a quick question - is this basically just the same PR as before, but now the original bug that we tripped has been fixed? If so, and if you've already tested it out, I'd be +1 on merging unless you want fresh eyes on any new stuff in particular, since the last PR already had some approves

File: config/hubs/schema.yaml [hunk 2]
```diff
@@ -198,8 +206,13 @@ properties:
                       "<clustername>-<hubname>".
 
                       See https://manage.auth0.com/dashboard/us/2i2c/connections/database.
-            required:
-              - connection
+            if:
```
Comment (yuvipanda) [inline config/hubs/schema.yaml:209]: this is super cool, TIL! Thanks @damianavila!

File: deployer/hub.py [hunk 3]
```diff
@@ -348,6 +348,7 @@ def get_generated_config(self, auth_provider: KeyProvider, secret_key):
             # FIXME: We're hardcoding Auth0OAuthenticator here
```
Comment (yuvipanda) [inline deployer/hub.py:348]: Update the comment to match the code change?

**Review by yuvipanda — CHANGES_REQUESTED** [review]
A code comment doesn't match what the code is doing now, so we should just update that? Otherwise this is good to me.

**Comment by sgibson91** [comment]
I've updated the comment @yuvipanda pointed out, so I will merge once tests pass :)

File: config/hubs/schema.yaml (hunk 2 above)
Comment (damianavila) [reply config/hubs/schema.yaml:209]: Yeah... I really liked it when I discovered it 😉

**Pull request merged**